Dry run (--check mode) prints discrete banner at the beginning and the end of the playbook, play and task execution.

basos9 opened the pull request:
##### SUMMARY
fixes #48787
<!--- Describe the change below, including rationale and design decisions -->
Dry run (--check mode) should print discrete banner at the beginning and more importantly at the end of the playbook execution, in order to clarify understanding of what was actually done.

When running long playbooks it might be good practice to first execute a dry run in order to have an idea of the actual changes that will be made. Also one could accompany this mode with --diff in order to see file changes and such, for modules that support this.

In the current implementation the output of a normal (wet) and dry run is the same. Thus, it can become very confusing what was the the actual mode of a playbook when scrolling back at the playbook output, or even worse when reading a log file, if the output is logged (via ansible configuration option log_path). In this case it is impossible to understand if the run was a dry or a wet one.

Thus it is proposed to prepend and append a banner (this one being more important since it is viewed next to the playbook summary) when a playbook is run in check mode (dry run). This should also be printed to the log file.

Also after discussions in #48873, this adds discrete banners to the PLAY and TASK indicating the play's or task's check mode either from option configuration (`check_mode: yes` option) or from runtime (`--dry-run` argument). 

The difference in TASK semantics is that check_mode can be enabled from task config, meaning that a module is configured to execute in check mode in order to achieve the normal playbook outcome (wet run without --check-mode) (possilby having the purpose to collect/display information).

On the other hand PLAYBOOK runtime argument, means that the operator/administrator of the playbook needs to invoke a dry run. It is up to the playbook to implement this correctly. Usually in this mode, most (if not all) of the tasks, will be executed in CHECK MODE in order to implement the PLAYBOOK's dry run.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ansible

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
$ ansible-playbook -C -i hosts test/testinc.yml 

DRY RUN ****************************************************************************************************************************

PLAY [all] [CHECK MODE] ************************************************************************************************************

TASK [Something in the way] [CHECK MODE] *******************************************************************************************
ok: [localhost] => {
    "msg": "Inside include playbook"
}

PLAY RECAP *************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0   


DRY RUN ****************************************************************************************************************************
```

basos9 committed:
ADD check mode markers feature with configuration option (default off)
ADD DRY RUN marker at the beggining and ending of pb execution
ADD CHECK MODE marker at the play and task output

bcoca commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -213,7 +215,8 @@ def _print_task_banner(self, task):
             task_name = task.get_name().strip()
 
-        self._display.banner(u"%s [%s%s]" % (prefix, task_name, args))
+        checkmsg = " [CHECK MODE]" if task.check_mode or (self._play_context and self._play_context.check_mode) else ""
```
i would not depend on play_context, the play should reflect the correct check_mode for itself at that point

basos9 replied:
There is an issue that I fixed in a new force commit. It shoud read
```
        checkmsg = " [CHECK MODE]" if task.check_mode or (task.check_mode is None and self._play_context and self._play_context.check_mode) else ""
```

The intended and checked behavior is :
- When the play has a defined check_mode this applies. 
- Else the runtime config from _play_context applies

This is kind of duplication of logic of task options and playcontext runtime arguments merging, that is implemented somewhere in deep in TaskExectutor. 

Also, as I checked with Pdb `task.check_mode` is the task's option config and not the runtime config overridden when applicable.

bcoca commented on lib/ansible/executor/playbook_executor.py:
```diff
@@ -75,7 +75,7 @@ def run(self):
         try:
             for playbook_path in self._playbooks:
-                pb = Playbook.load(playbook_path, variable_manager=self._variable_manager, loader=self._loader)
+                pb = Playbook.load(playbook_path, variable_manager=self._variable_manager, loader=self._loader, executor=self)
```
no need to pass playbook_executor, callback already has the cli options/config and can check if it was run in check_mode `self._options['check_mode']`

basos9 replied:
Unfortunately this is not the case (as checked with pdb).
Possibly the cause is the way callbacks are loaded in 

`lib/ansible/executor/task_queue_manager.py`

```
                    self._stdout_callback.set_options()
```

The other path from callback base seems to not work, as cli is None

`lib/ansible/plugins/callback/__init__.py`

```
        if cli:
            self._options = cli.options
        else:
```

Indeed I didn't like to induce the PBExecutor->Playbook coupling for only to access a variable, but this was the least intervening method. The other option that I considered and disregarded was to add a new callback in order to set the PBExecutor (too much hustle). And last consideration was to modify the on_playbook_start signature but this also would involve changes to many existing callback signatures.

bcoca replied:
'cli' should be loaded, otherwise that is a bug, the conditional there is for those using the API directly. the `set_options` there refers to plugin_options not 'options' (sorry for name overload, but one is general across plugins and i forgot callbacks already were using it for cli options.

in any case i'm in the process of removing play_context from option handling so it is not a good idea to rely on it

basos9 replied:
Regarding your second though (that play_context would be decoupled from options) 
- We could either keep somehow the link between playbook Executor and callback (in order to access run time option) and disregard the _play_context. Indeed I could have used playbook executor for both the CHECK MODE message code above and the DRY RUN one.

- Or (taking into account the second your first observation) the None cli bug should be investigated. I think I should leave this to some more experienced with ansible internals.

basos9 replied:
Now the refs to playbookExecutor and play_context have been removed, and we use the newly introduced context.

basos9 commented:
## TESTING playbook
Run with and without --check-mode

```yaml
---
  - hosts: all
    gather_facts: no
    tasks:

    - name: Somehting in a way
      debug: 
        msg: 'Inside {{ descr | default("") }} playbook (check {{ansible_check_mode}})'

    - name: Command (NOCHK)
      command: ls -l
      check_mode: false

    - name: Command(CHK)
      command: ls -l
      check_mode: true

    - name: Command
      command: ls -l
```

#### Execution with --check-mode
```

DRY RUN ******************************************************************************************

TASK [Somehting in a way] [CHECK MODE] ***********************************************************
ok: [localhost] => {
    "msg": "Inside  playbook (check True)"
}

TASK [Command (NOCHK)] ***************************************************************************
changed: [localhost]

TASK [Command(CHK)] [CHECK MODE] *****************************************************************
skipping: [localhost]

TASK [Command] [CHECK MODE] **********************************************************************
skipping: [localhost]

PLAY RECAP ***************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=2   
```

#### Execution without checkmode
```
TASK [Somehting in a way] ************************************************************************
ok: [localhost] => {
    "msg": "Inside  playbook (check False)"
}

TASK [Command (NOCHK)] ***************************************************************************
changed: [localhost]

TASK [Command(CHK)] [CHECK MODE] *****************************************************************
skipping: [localhost]

TASK [Command] ***********************************************************************************
changed: [localhost]

PLAY RECAP ***************************************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0    skipped=1   


```

basos9 commented:
@bcoca Hello, any moving thoughts for this ?

nqb commented:
Would be nice to have in 2.7.6

basos9 commented:
Now the refs to playbookExecutor and play_context have been removed, and we use the newly introduced context.

agaffney commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -214,7 +214,8 @@ def _print_task_banner(self, task):
             task_name = task.get_name().strip()
 
-        self._display.banner(u"%s [%s%s]" % (prefix, task_name, args))
+        checkmsg = " [CHECK MODE]" if task.check_mode or (task.check_mode is None and context.CLIARGS['check']) else ""
```
Just for clarity when reading the code, this should probably be broken out into multiple lines. It's a bit too long for a simple ternary (imo).

basos9 replied:
Would this be more appropriate ?
```
if task.check_mode or (task.check_mode is None and context.CLIARGS['check']):
  checkmsg = " [CHECK MODE]"  
else 
  checkmsg = ""
```


I prefer the compact one line form for such a simple assignment statement. Also there are other long lines e.g.
```
if (self._display.verbosity > 2 or '_ansible_verbose_always' in result._result) and '_ansible_verbose_override' not in result._result:
```

agaffney replied:
Yes, that would be preferable.

That other long line you pointed out is *just* if the `if` condition, not the entire if/else construct like the line I pointed out here.

basos9 replied:
Ok, I get yr point.

bcoca reviewed (changes requested):
I would also add a configuration toggle for the callback, since we don't want to break existing setups by changing the default output 'check_mode_prefix: false|true'

basos9 commented:
Seems logical.

Do you mean a ansible.cfg type option ? 
like
```
check_mode_prefix: false|true
```
and ENV
```
CHECK_MODE_PREFIX=1 
```

bcoca commented:
you can add that (and env var) in the DOCUMENTATION section of the plugin

basos9 commented:
Added config option, with this name, as it seems more descriptive.
```
CHECK_MODE_MARKERS=0

[defaults]
check_mode_markers=0
```

I was not sure where to put documentation. I added a base config section and a doc_fragments (not sure where this is used). Feel free to comment.

basos9 commented:
Finaly, as I understood, the new options are added only as plugin options and not base options. I hope this should be ok.

basos9 commented:
@bocoa hello, just a reminder that this is fixed. If there is time please check it, thanks.

bcoca commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -36,7 +36,8 @@
                   ('show_custom_stats', C.SHOW_CUSTOM_STATS),
-                  ('display_failed_stderr', False),)
+                  ('display_failed_stderr', False),
+                  ('check_mode_markers', True),)
```
why are you adding to compat options? this is new option

basos9 replied:
I did what is done for `display_ok_hosts` which is a new type option. As i understand the alternative would be to add
```
        self.check_mode_markers = self.get_option('check_mode_markers')
```
in set_options()

But wouldn't this cause issues regarding the comment about old plugins ?
```
# These values use ansible.constants for historical reasons, mostly to allow
# unmodified derivative plugins to work.  
.......
```

bcoca commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -213,7 +214,11 @@ def _print_task_banner(self, task):
             task_name = task.get_name().strip()
 
-        self._display.banner(u"%s [%s%s]" % (prefix, task_name, args))
+        if (task.check_mode or (task.check_mode is None and context.CLIARGS['check'])) and self.check_mode_markers:
```
dont need to check CLIARGs as task check_mode should already reflect this

basos9 replied:
Yes, indeed this is not needed any more. Fixed.

basos9 commented:
@bcoca hello, this is fixed in my opinion. Please add any comments. Thanks

jamescassell commented on lib/ansible/plugins/doc_fragments/default_callback.py:
```diff
@@ -55,6 +55,19 @@ class ModuleDocFragment(object):
+        type: bool
+        default: yes
+        env:
+          - name: CHECK_MODE_MARKERS
```
```suggestion
          - name: ANSIBLE_CHECK_MODE_MARKERS
```

basos9 replied:
done

jamescassell reviewed (changes requested):
All ansible-defined environment variables should start with `ANSIBLE_`

basos9 commented:
#### PLUGINS THAT INHERIT FROM DEFAULT AND USE DOC FRAGMENT
```
full_skip.py    WORKS   DEPRECATED 
debug.py        WORKS
skippy.py       WORKS   DEPRECATED
stderr.py       WORKS   DEPRECATED
yaml.py         WORKS   
actionable.py   WORKS   DEPRECATED
dense.py        NOTOK (it does not output ending banner)
```

#### PLUGINS THAT ONLY USE DOC FRAGMENT
counter_enabled.py
unixy.py

basos9 commented:
A full scenario testing playbook, run with and without -C argument. Tests all cases of playbook, play and task check mode settings. 

```
## VARS:
#  descr
---
  - hosts: all
    gather_facts: no
    tasks:

    - name: Somehting in a way
      debug: 
        msg: 'Inside {{ descr | default("") }} playbook (check {{ansible_check_mode}})'

    - name: Command (NOCHK)
      command: ls -l
      check_mode: false

    - name: Command(CHK)
      command: ls -l
      check_mode: true

    - name: Command
      command: ls -l


  - hosts: all
    gather_facts: no
    check_mode: true
    tasks:

    - name: Somehting in a way (P:CHK)
      debug: 
        msg: 'Inside {{ descr | default("") }} playbook (check {{ansible_check_mode}})'

    - name: Command (NOCHK) (P:CHK)
      command: ls -l
      check_mode: false

    - name: Command(CHK) (P:CHK)
      command: ls -l
      check_mode: true

  - hosts: all
    gather_facts: no
    check_mode: false
    tasks:

    - name: Somehting in a way (P:NOCHK)
      debug: 
        msg: 'Inside {{ descr | default("") }} playbook (check {{ansible_check_mode}})'

    - name: Command (NOCHK) (P:NOCHK)
      command: ls -l
      check_mode: false

    - name: Command(CHK) (P:NOCHK)
      command: ls -l
      check_mode: true
```

bcoca commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,25 @@
+        at the beggining and ending of playbooks execution (when calling ansible-playbook --check)
+        and 'CHECK MODE' as a suffix at every task that is run in 'check' mode."
+        type: bool
+        default: yes
```
default should be 'no' to keep current behaviout

also missing `version_added: "2.9"`

basos9 replied:
Default changed to no as per our discussions.
To be enabled after a couple of majors (after 2.9)

basos9 commented:
Hello @bcoca. Is there anything else to discuss for this?

bcoca commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,25 @@
+        at the beggining and ending of playbooks execution (when calling ansible-playbook --check)
+        and 'CHECK MODE' as a suffix at every task that is run in 'check' mode."
+        type: bool
+        default: yes
```
version_added: '2.9'

basos9 replied:
this was ok until a wrong push of me  (789620b to 6ed2867 3 hours ago), which is now again fixed (with the push  6ed2867 to 255f5c3 4 minutes ago)

basos9 commented:
#### Integration test
```<bcoca> current output unchagned and 'expected' results in diff scenarios x) --check-mode x) check_mode:true on task/block/role
<sdoran> zer0_her0: Look at test/integration/targets/callback_default/runme.sh
<jtanner> new folder like tests/integration/targets/dryrun, with a runme.sh file in there that executes the playbook with the new feature and checks that it worked
<zer0_her0> ok, I'll c heck this
<sdoran> That should give you an idea of what to do. You can ask for further help in #ansible-devel if you get stuck.
<jtanner> and on the the ansible-devel mailing list````

basos9 committed:
ADD tests for default callback check mode markers

basos9 commented:
Integration **tests** added

samdoran commented on changelogs/fragments/check_mode_markers.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+    - Added check mode markers (DRY RUN, CHECK_MODE) when running a pb in check mode. By default this is off until tested, use ANSIBLE_CHECK_MODE_MARKERS=1 to enable it.
```
```suggestion
    - Add new option to default standard out callback plugin, ``ANSIBLE_CHECK_MODE_MARKERS ``, which adds check mode markers (``DRY RUN``, ``CHECK_MODE``) to the output when running in check mode. It is off by default.
```

basos9 replied:
ok

samdoran commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,27 @@
+      check_mode_markers:
+        name: Show markers when running on check mode
+        description:
+        - "Toggle to control displaying markers when running on check mode. The markers are 'DRY RUN'
```
```suggestion
        - "Toggle to control displaying markers when running in check mode. The markers are C(DRY RUN)
```

samdoran commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,27 @@
+        name: Show markers when running on check mode
+        description:
+        - "Toggle to control displaying markers when running on check mode. The markers are 'DRY RUN'
+        at the beggining and ending of playbooks execution (when calling ansible-playbook --check)
```
```suggestion
        at the beggining and ending of play execution (when calling C(ansible-playbook --check))
```

basos9 replied:
Actually I meant playbook here. See example run in the[ first comment](https://github.com/ansible/ansible/pull/49432#issue-235408755). The full corrected message is 

> The markers are C(DRY RUN) at the beggining and ending of playbook execution (when calling C(ansible-playbook --check)) and C(CHECK MODE) as a suffix at every play and task that is run in check mode."

samdoran commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,27 @@
+        description:
+        - "Toggle to control displaying markers when running on check mode. The markers are 'DRY RUN'
+        at the beggining and ending of playbooks execution (when calling ansible-playbook --check)
+        and 'CHECK MODE' as a suffix at every task that is run in 'check' mode."
```
```suggestion
        and C(CHECK MODE) as a suffix at every task that is run in check mode."
```

samdoran commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,27 @@
+            section: defaults
 '''
 
+# NOTE: check_mode_markers functionality is also implemented in the following derived plugins:
```
The code is implemented there but with the options only existing in the default callback plugin, there is no way to enable it with other callback plugins. 

I would suggest removing this note and moving the `check_mode_markers` documentation to the default doc fragment (`lib/ansible/plugins/doc_fragments/default_callback.py`). This will allow the setting to be controlled from all those plugins as well as update the documentation for all those plugins.

basos9 replied:
Hello, please have a look at [comment](https://github.com/ansible/ansible/pull/49432#issuecomment-485834431) above for clarification, regarding a discussion we had with bcoca about this.

In short, there are plugins which use the default callback's doc fragment but do not inherit from it (like others).

So in the present situation of this patch, where documentation is in the default plugin, there is an undocumented feature for the three mentioned plugins. 

On the other hand if I move the documentation to the default callback's doc fragment, there will be 2 plugins that will have their documentation mention the feature but in fact do not implement it.

samdoran replied:
Ok. I did not see that previous discussion.

samdoran commented on lib/ansible/plugins/callback/default.py:
```diff
@@ -16,8 +16,27 @@
 
+# NOTE: check_mode_markers functionality is also implemented in the following derived plugins:
+#       debug.py, yaml.py, dense.py. Maybe their documentation needs updating, too.
+# TODO: check_mode markets make default after some stable releases
```
```suggestion
# TODO: Change the default to True in a future release (2.13)
```
